File: server/core/taskProtocol.js
const fs = require('fs');
const path = require('path');

/**
 * 任务对象协议
 *
 * 统一 Node 和 Python 之间的任务通信格式，避免文件名耦合。
 *
 * 文件结构：
 * - task.json: 任务输入（Node 写入，Python 读取）
 * - result.json: 任务成功输出（Python 写入，Node 读取）
 * - failure.json: 任务失败输出（Python 写入，Node 读取）
 */

/**
 * 创建任务输入对象
 * @param {string} taskId - 任务 ID
 * @param {string} type - 任务类型（vertical_queue, pipeline, xai_top10, wechat_rpa）
 * @param {object} input - 任务输入参数
 * @param {string} workDir - 工作目录
 * @returns {object} 任务输入对象
 */
function createTaskInput(taskId, type, input, workDir) {
  return {
    taskId,
    type,
    input: input || {},
    workDir: workDir || '',
    createdAt: new Date().toISOString()
  };
}

/**
 * 创建任务成功输出对象
 * @param {string} taskId - 任务 ID
 * @param {object} artifacts - 产物清单（文件名映射）
 * @param {object} metadata - 元数据
 * @returns {object} 任务成功输出对象
 */
function createTaskResult(taskId, artifacts, metadata) {
  return {
    taskId,
    status: 'success',
    artifacts: artifacts || {},
    metadata: metadata || {},
    completedAt: new Date().toISOString()
  };
}

/**
 * 创建任务失败输出对象
 * @param {string} taskId - 任务 ID
 * @param {object} error - 错误信息
 * @returns {object} 任务失败输出对象
 */
function createTaskFailure(taskId, error) {
  return {
    taskId,
    status: 'failed',
    error: {
      code: error.code || 'UNKNOWN_ERROR',
      message: error.message || 'Unknown error',
      stage: error.stage || 'unknown',
      details: error.details || ''
    },
    failedAt: new Date().toISOString()
  };
}

/**
 * 写入任务输入文件
 * @param {string} workDir - 工作目录
 * @param {object} taskInput - 任务输入对象
 */
function writeTaskInput(workDir, taskInput) {
  const taskPath = path.join(workDir, 'task.json');
  fs.writeFileSync(taskPath, JSON.stringify(taskInput, null, 2), 'utf-8');
}

/**
 * 读取任务输入文件
 * @param {string} workDir - 工作目录
 * @returns {object|null} 任务输入对象，不存在返回 null
 */
function readTaskInput(workDir) {
  const taskPath = path.join(workDir, 'task.json');
  if (!fs.existsSync(taskPath)) {
    return null;
  }
  try {
    return JSON.parse(fs.readFileSync(taskPath, 'utf-8'));
  } catch (_err) {
    return null;
  }
}

/**
 * 写入任务结果文件
 * @param {string} workDir - 工作目录
 * @param {object} taskResult - 任务结果对象
 */
function writeTaskResult(workDir, taskResult) {
  const resultPath = path.join(workDir, 'result.json');
  fs.writeFileSync(resultPath, JSON.stringify(taskResult, null, 2), 'utf-8');
}

/**
 * 读取任务结果文件
 * @param {string} workDir - 工作目录
 * @returns {object|null} 任务结果对象，不存在返回 null
 */
function readTaskResult(workDir) {
  const resultPath = path.join(workDir, 'result.json');
  if (!fs.existsSync(resultPath)) {
    return null;
  }
  try {
    return JSON.parse(fs.readFileSync(resultPath, 'utf-8'));
  } catch (_err) {
    return null;
  }
}

/**
 * 写入任务失败文件
 * @param {string} workDir - 工作目录
 * @param {object} taskFailure - 任务失败对象
 */
function writeTaskFailure(workDir, taskFailure) {
  const failurePath = path.join(workDir, 'failure.json');
  fs.writeFileSync(failurePath, JSON.stringify(taskFailure, null, 2), 'utf-8');
}

/**
 * 读取任务失败文件
 * @param {string} workDir - 工作目录
 * @returns {object|null} 任务失败对象，不存在返回 null
 */
function readTaskFailure(workDir) {
  const failurePath = path.join(workDir, 'failure.json');
  if (!fs.existsSync(failurePath)) {
    return null;
  }
  try {
    return JSON.parse(fs.readFileSync(failurePath, 'utf-8'));
  } catch (_err) {
    return null;
  }
}

/**
 * 读取任务输出（优先读取 result.json，回退到 failure.json）
 * @param {string} workDir - 工作目录
 * @returns {object|null} 任务输出对象，不存在返回 null
 */
function readTaskOutput(workDir) {
  const result = readTaskResult(workDir);
  if (result) {
    return result;
  }
  return readTaskFailure(workDir);
}

/**
 * 检查任务是否完成（成功或失败）
 * @param {string} workDir - 工作目录
 * @returns {boolean} 是否完成
 */
function isTaskCompleted(workDir) {
  return fs.existsSync(path.join(workDir, 'result.json')) ||
         fs.existsSync(path.join(workDir, 'failure.json'));
}

/**
 * 解析产物路径（相对路径转绝对路径）
 * @param {string} workDir - 工作目录
 * @param {object} artifacts - 产物清单
 * @returns {object} 绝对路径的产物清单
 */
function resolveArtifactPaths(workDir, artifacts) {
  const resolved = {};
  for (const [key, value] of Object.entries(artifacts || {})) {
    if (typeof value === 'string') {
      resolved[key] = path.isAbsolute(value) ? value : path.join(workDir, value);
    } else {
      resolved[key] = value;
    }
  }
  return resolved;
}

module.exports = {
  createTaskInput,
  createTaskResult,
  createTaskFailure,
  writeTaskInput,
  readTaskInput,
  writeTaskResult,
  readTaskResult,
  writeTaskFailure,
  readTaskFailure,
  readTaskOutput,
  isTaskCompleted,
  resolveArtifactPaths
};

File: python/pipeline/task_protocol.py
"""
任务对象协议 - Python 端工具模块

统一 Node 和 Python 之间的任务通信格式，避免文件名耦合。
"""

import json
import os
from datetime import datetime
from typing import Dict, Any, Optional


def read_task_input(work_dir: str) -> Optional[Dict[str, Any]]:
    """
    读取任务输入文件

    Args:
        work_dir: 工作目录

    Returns:
        任务输入对象，不存在返回 None
    """
    task_path = os.path.join(work_dir, 'task.json')
    if not os.path.exists(task_path):
        return None

    try:
        with open(task_path, 'r', encoding='utf-8') as f:
            return json.load(f)
    except Exception:
        return None


def write_task_result(work_dir: str, task_id: str, artifacts: Dict[str, Any], metadata: Dict[str, Any] = None):
    """
    写入任务成功输出文件

    Args:
        work_dir: 工作目录
        task_id: 任务 ID
        artifacts: 产物清单（文件名映射）
        metadata: 元数据
    """
    result = {
        'taskId': task_id,
        'status': 'success',
        'artifacts': artifacts or {},
        'metadata': metadata or {},
        'completedAt': datetime.utcnow().isoformat() + 'Z'
    }

    result_path = os.path.join(work_dir, 'result.json')
    with open(result_path, 'w', encoding='utf-8') as f:
        json.dump(result, f, indent=2, ensure_ascii=False)


def write_task_failure(work_dir: str, task_id: str, error_code: str, error_message: str,
                       error_stage: str = 'unknown', error_details: str = ''):
    """
    写入任务失败输出文件

    Args:
        work_dir: 工作目录
        task_id: 任务 ID
        error_code: 错误码
        error_message: 错误消息
        error_stage: 错误阶段
        error_details: 错误详情
    """
    failure = {
        'taskId': task_id,
        'status': 'failed',
        'error': {
            'code': error_code,
            'message': error_message,
            'stage': error_stage,
            'details': error_details
        },
        'failedAt': datetime.utcnow().isoformat() + 'Z'
    }

    failure_path = os.path.join(work_dir, 'failure.json')
    with open(failure_path, 'w', encoding='utf-8') as f:
        json.dump(failure, f, indent=2, ensure_ascii=False)


def get_artifact_path(work_dir: str, artifact_name: str) -> str:
    """
    获取产物的绝对路径

    Args:
        work_dir: 工作目录
        artifact_name: 产物文件名（相对路径）

    Returns:
        产物的绝对路径
    """
    if os.path.isabs(artifact_name):
        return artifact_name
    return os.path.join(work_dir, artifact_name)
